File: algoritmo_primaria.c
#include <stdio.h>
#include <string.h>

#define MAX 1000  // dígitos máximos

// Imprime un arreglo de dígitos alineado a la derecha (sin ceros a la izquierda)
void print_digits_right(const int *a, int n) {
    int i = 0;
    while (i < n - 1 && a[i] == 0) i++;  // quitar ceros iniciales
    // espacios a la izquierda para alinear a n columnas
    for (int s = 0; s < i; s++) putchar(' ');
    for (; i < n; i++) printf("%d", a[i]);
    putchar('\n');
}

void print_line(int n) {
    for (int i = 0; i < n; i++) putchar('-');
    putchar('\n');//incluye una linea 
}

int main(void) {
    char num1[MAX], num2[MAX];
    printf("Ingrese el primer numero: ");
    scanf("%s", num1);
    printf("Ingrese el segundo numero: ");
    scanf("%s", num2);

    int len1 = (int)strlen(num1);
    int len2 = (int)strlen(num2);
    int width = len1 + len2;            // ancho máximo (dígitos del resultado)
    int resultado[MAX] = {0};           // acumulador final (right-aligned en width)

    // Encabezado (alineado a la derecha)
    printf("\n");
    printf("%*s\n", width, num1);
    printf("%*s%s\n", width - len2 - 2, "", "x ");
    printf("%*s\n", width, num2);
    print_line(width);

    // Para cada dígito del multiplicador (num2), de derecha a izquierda
    for (int j = len2 - 1; j >= 0; j--) {
        int d2 = num2[j] - '0';
        int parcial[MAX] = {0};

        // desplazamiento: cuántos ceros a la derecha (como en el cuaderno)
        int shift = (len2 - 1) - j;

        // índice de escritura empezando desde la derecha,
        // dejando 'shift' posiciones libres al final
        int k = width - 1 - shift;
        int carry = 0;

        // multiplicar num1 (de derecha a izquierda) por d2
        for (int i = len1 - 1; i >= 0; i--) {
            int d1 = num1[i] - '0';
            int prod = d1 * d2 + carry;
            parcial[k] = prod % 10;   // escribir dígito
            carry = prod / 10;
            k--;                      // moverse hacia la izquierda
        }
        if (carry > 0 && k >= 0) {
            parcial[k] = carry;
        }

        // Imprimir fila parcial (alineada, sin 0 extra a la derecha)
        print_digits_right(parcial, width);

        // Sumar el parcial al resultado
        int c = 0;
        for (int p = width - 1; p >= 0; p--) {
            int s = resultado[p] + parcial[p] + c;
            resultado[p] = s % 10;
            c = s / 10;
        }
    }

    print_line(width);
    printf("Resultado: ");
    // Imprimir resultado final sin ceros a la izquierda
    int start = 0;
    while (start < width - 1 && resultado[start] == 0) start++;
    for (int i = start; i < width; i++) printf("%d", resultado[i]);
    printf("\n");

    return 0;
}

File: algoritmo_primaria.py
def multiplicar_con_proceso(a: str, b: str) -> str:
    # --- Normalización y signo ---
    a = a.strip()
    b = b.strip()
    signo = 1
    if a.startswith('-'):
        signo *= -1
        a = a[1:]
    if b.startswith('-'):
        signo *= -1
        b = b[1:]
    a = a.lstrip('0') or '0'
    b = b.lstrip('0') or '0'

    # Caso directo si alguno es 0
    if a == '0' or b == '0':
        width = max(len(a) + len(b), 1)
        print()
        print(a.rjust(width))
        print(('x ' + b).rjust(width))
        print('-' * width)
        print('0'.rjust(width))
        print('-' * width)
        print('Resultado:', '0')
        return '0'

    len1, len2 = len(a), len(b)
    width = len1 + len2  # máximo de dígitos del producto
    resultado = [0] * width

    # --- Encabezado alineado a la derecha ---
    print()
    print(a.rjust(width))
    print(('x ' + b).rjust(width))
    print('-' * width)

    # --- Filas parciales (de derecha a izquierda en el multiplicador) ---
    for j in range(len2 - 1, -1, -1):
        d2 = ord(b[j]) - 48
        parcial = [0] * width
        shift = (len2 - 1) - j            # cuántos "ceros" a la derecha
        k = width - 1 - shift             # empezamos a escribir desde la derecha menos el shift
        carry = 0

        for i in range(len1 - 1, -1, -1):
            d1 = ord(a[i]) - 48
            prod = d1 * d2 + carry
            parcial[k] = prod % 10
            carry = prod // 10
            k -= 1
        if carry > 0 and k >= 0:
            parcial[k] = carry

        # imprimir fila parcial (sin ceros a la izquierda, pero alineada a 'width')
        s_parcial = ''.join(map(str, parcial)).lstrip('0') or '0'
        print(s_parcial.rjust(width))

        # sumar parcial al acumulado
        c = 0
        for p in range(width - 1, -1, -1):
            s = resultado[p] + parcial[p] + c
            resultado[p] = s % 10
            c = s // 10

    # --- Resultado final ---
    print('-' * width)
    res = ''.join(map(str, resultado)).lstrip('0') or '0'
    if signo < 0 and res != '0':
        res = '-' + res
    print('Resultado:', res)
    return res


# --- Ejecución interactiva ---
if __name__ == "__main__":
    n1 = input("Ingrese el primer numero: ")
    n2 = input("Ingrese el segundo numero: ")
    multiplicar_con_proceso(n1, n2)
